embulk-core/src/main/java/org/embulk/config/DataSourceImpl.java

```diff
@@ -128,28 +128,40 @@ public DataSourceImpl set(String attrName, Object v) {
         return this;
     }
 
-    @SuppressWarnings("deprecation")  // Calling getObjectNode().
     @Override
     public DataSourceImpl setNested(String attrName, DataSource v) {
-        if (v instanceof DataSourceImpl) {
-            final DataSourceImpl vImpl = (DataSourceImpl) v;
-            data.set(attrName, vImpl.getObjectNode());
+        if (v == null) {
+            this.data.set(attrName, null);
         } else {
-            throw new IllegalArgumentException("DataSourceImpl#setNested aceepts only DataSourceImpl.");
+            final String vJsonStringified = v.toJson();
+            if (vJsonStringified == null) {
+                throw new ConfigException(new NullPointerException("DataSource#setNested accepts only valid DataSource"));
+            }
+            final JsonNode vJsonNode = this.model.readObject(JsonNode.class, vJsonStringified);
+            if (!vJsonNode.isObject()) {
+                throw new ConfigException(new ClassCastException("DataSource#setNested accepts only valid JSON object"));
+            }
+            this.data.set(attrName, (ObjectNode) vJsonNode);
```

Comment:
We don't need this cast since checked if it's `JsonObject`.

Reply:
Ah, that's true. `ObjectNode#set` receives `JsonNode`.
https://fasterxml.github.io/jackson-databind/javadoc/2.6/com/fasterxml/jackson/databind/node/ObjectNode.html#set-java.lang.String-com.fasterxml.jackson.databind.JsonNode-

But I think it's okay to keep it as it has no cost, and we can explicitly declare it should be `ObjectNode` here. If you think it's to be removed, please let me know.